src/main/kotlin/com/dp/advancedgunnerycontrol/weaponais/AGCUitls.kt

```diff
@@ -243,3 +244,35 @@ fun mapBooleanToSpecificString(boolValue: Boolean, trueString: String, falseStri
 internal infix fun Vector2f.times_(d: Float): Vector2f {
     return Vector2f(d * x, d * y)
 }
+
+
+/**
+ * @return approximate angular distance of target from current weapon facing in rad
+ * note: approximation works well for small values and is off by a factor of PI/2 for 180°
+ * @param entity: Relative coordinates (velocity-compensated)
+ */
+fun angularDistanceFromWeapon(entity: Vector2f, weapon: WeaponAPI): Float {
+    val weaponDirection = vectorFromAngleDeg(weapon.currAngle)
+    val distance = entity - weapon.location
+    val entityDirection = distance times_ (1f / distance.length())
+    return (weaponDirection - entityDirection).length()
+}
+
+fun linearDistanceFromWeapon(entity: Vector2f, weapon: WeaponAPI): Float {
+    return (weapon.location - entity).length()
+}
+/**
```

Comment:
I arrived at the strict method myself, so the explanation will be wonky at best. Anyway, it works the following way. First, we design a function `f(x)=distSquared(aimPoint * x, entity)` where `aimPoint` and `entity` are coordinates relative to weapon location. We want to find the minimum of `f(x)`. If it's below `collRadius^2` then we have a hit (except maybe some edge cases). To do that, we calculate `d/dx f(x)` and solve it for `x`. Then we simply use the solution to get the min distance. I tested the code separately, and didn't find any problems. Though it may fail eg when `aimPoint` is same as wapon location.

Reply:
Wait a moment, one does not need all the derivative stuff to find quadratic function minimum. I guess I rediscovered high-school level math. Lame.

Reply:
I think I remember a linear algebra class from university where our professor showed us some proof that a straight orthogonal line is guaranteed to be the shortest path. Then he said "luckily, humans and animals know this instinctively, otherwise we would have all gone extinct a long time ago" xD

Overall, I'm pretty sure that simply aiming for the point where the enemy is predicted to be in the time it takes the projectile to travel to that point, taking the current momentum of the firing ship into account*, is gonna be the optimal solution. At least that's what my human instinct tells me :P

*for the first implementation of my custom weapon AI I failed to account for the firing ship's momentum. A "clean" solution would probably have been to have a function/class that transforms between global and relative coordinates, but that seemed like lots of work so instead I added the compensateTargetPointShipSpeed method and added a comment to the top of the class (SpecificAiPluginBase)^^

Reply:
The tricky part is predicting where the target will be when the projectile reaches it! The time of flight depends on the predicted location and the predicted location depends on the time of flight.